File: scenarios/vuln_scan/fixtures/scanner.py
from __future__ import annotations

import json
import sys
import urllib.request
from pathlib import Path


def version_tuple(value: str) -> tuple[int, ...]:
    return tuple(int(part) for part in value.split("."))


def main() -> None:
    manifest_path = Path(sys.argv[1])
    advisory_url = sys.argv[2]
    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    with urllib.request.urlopen(advisory_url, timeout=10) as response:
        advisories = json.loads(response.read().decode("utf-8"))["advisories"]

    findings = []
    for dep in manifest["dependencies"]:
        for advisory in advisories:
            if dep["name"] != advisory["package"]:
                continue
            if version_tuple(dep["version"]) < version_tuple(advisory["affected_below"]):
                findings.append(
                    {
                        "id": advisory["id"],
                        "package": dep["name"],
                        "version": dep["version"],
                        "severity": advisory["severity"],
                    }
                )
    payload = {
        "findings": findings,
        "summary": {
            "total": len(findings),
            "high": sum(1 for item in findings if item["severity"] == "high"),
            "medium": sum(1 for item in findings if item["severity"] == "medium")
        }
    }
    print(json.dumps(payload))


if __name__ == "__main__":
    main()

File: scenarios/vuln_scan/fixtures/scanner.ts
import { readFile } from "node:fs/promises";

type Dependency = { name: string; version: string };
type Advisory = { package: string; affected_below: string; id: string; severity: string };

function versionTuple(value: string): number[] {
  return value.split(".").map((part) => Number(part));
}

function versionLessThan(left: string, right: string): boolean {
  const a = versionTuple(left);
  const b = versionTuple(right);
  const length = Math.max(a.length, b.length);
  for (let index = 0; index < length; index += 1) {
    const diff = (a[index] ?? 0) - (b[index] ?? 0);
    if (diff !== 0) {
      return diff < 0;
    }
  }
  return false;
}

const manifestPath = process.argv[2];
const advisoryUrl = process.argv[3];
if (!manifestPath || !advisoryUrl) {
  throw new Error("manifest path and advisory url are required");
}

const manifest = JSON.parse(await readFile(manifestPath, "utf8")) as { dependencies: Dependency[] };
const advisoryPayload = (await (await fetch(advisoryUrl)).json()) as { advisories: Advisory[] };

const findings = [];
for (const dep of manifest.dependencies) {
  for (const advisory of advisoryPayload.advisories) {
    if (dep.name !== advisory.package) {
      continue;
    }
    if (versionLessThan(dep.version, advisory.affected_below)) {
      findings.push({
        id: advisory.id,
        package: dep.name,
        version: dep.version,
        severity: advisory.severity,
      });
    }
  }
}

console.log(
  JSON.stringify({
    findings,
    summary: {
      total: findings.length,
      high: findings.filter((item) => item.severity === "high").length,
      medium: findings.filter((item) => item.severity === "medium").length,
    },
  }),
);
